core/sigagg/sigagg.go

```diff
@@ -86,39 +86,84 @@ func (a *Aggregator) aggregate(ctx context.Context, pubkey core.PubKey, parSigs
 		return nil, errors.New("require threshold signatures")
 	}
 
-	// Get all partial signatures.
-	blsSigs := make(map[int]tbls.Signature)
+	// The following tables shows the layout of the data to be aggregated.
+	// Note that the signature aggregation will be performed for each individual column.
+	// +=============+============+============+============+============+
+	// | Data/Charon |   Data0    |   Data1    |   Data2    |   Data3    |
+	// +=============+============+============+============+============+
+	// | Charon 0    | ParSig[00] | ParSig[01] | ParSig[02] | ParSig[03] |
+	// 	+-------------+------------+------------+------------+------------+
+	// | Charon 1    | ParSig[10] | ParSig[01] | ParSig[12] | ParSig[13] |
+	// 	+-------------+------------+------------+------------+------------+
+	// | Charon 2    | ParSig[20] | ParSig[21] | ParSig[22] | ParSig[23] |
+	// 	+-------------+------------+------------+------------+------------+
+	// | Charon 3    | ParSig[30] | ParSig[31] | ParSig[32] | ParSig[33] |
+	// 	+-------------+------------+------------+------------+------------+
+	// | Aggregate   | Sig[0]     | Sig[1]     | Sig[2]     | Sig[3]     |
+	// 	+-------------+------------+------------+------------+------------+
+
+	parsigsPerPeer := make(map[int][]tbls.Signature)
 	for _, parSig := range parSigs {
-		sig, err := tblsconv.SigFromCore(parSig.Signature())
-		if err != nil {
-			return nil, errors.Wrap(err, "signature from core")
+		var sigs []tbls.Signature
+		for _, s := range parSig.Signatures() {
+			sig, err := tblsconv.SigFromCore(s)
+			if err != nil {
+				return nil, errors.Wrap(err, "signature from core")
+			}
+
+			sigs = append(sigs, sig)
 		}
-		blsSigs[parSig.ShareIdx] = sig
+
+		parsigsPerPeer[parSig.ShareIdx] = sigs
+	}
+
+	if len(parsigsPerPeer) < a.threshold {
+		return nil, errors.New("number of partial signatures less than threshold", z.Int("threshold", a.threshold), z.Int("got", len(parsigsPerPeer)))
 	}
 
-	if len(blsSigs) < a.threshold {
-		return nil, errors.New("number of partial signatures less than threshold", z.Int("threshold", a.threshold), z.Int("got", len(blsSigs)))
+	// Check if each peer has the same number of signatures.
+	sigLens := make(map[int]struct{})
+	var sigLen int
+	for _, row := range parsigsPerPeer {
+		sigLen = len(row)
+		sigLens[sigLen] = struct{}{}
+	}
+	if sigLen == 0 {
+		return nil, errors.New("empty signatures")
+	}
```

Comment:
this check can be moved to above for loop where `parsigsPerPeer` is set.
#failfast

Reply:
yeah done, please check update code